### PR TITLE
Uiuc vlm pr compressed fixed

### LINKED ISSUE #510: VLM support for DPO
DPO (and other SFT and RL methods) only supports text-only models as of now. VLM support will be very helpful.

**Checklist**

- [x ] I have searched the existing issues for similar feature requests.
- [x ] This is not a support question (please use the "bug template" for that).

### PR DESCRIPTION
**Pull request by immuntasir**
Resolves #510 

This PR introduces **multimodal support to Tunix’s Gemma3** model and **adds a new vision-language DPO demonstration notebook** (vl_dpo_demo_gemma3.ipynb), extending the framework to handle image-text reasoning and multimodal alignment.  Key changes includes:
- **Multimodal Gemma3 Model:**
    1. Added SigLIP vision encoder and MultimodalProjector to Gemma3, enabling joint image–text forward passes.
    2. Updated ModelConfig and parameter mapping to support multimodal checkpoints (multimodal=True).
- **DPO Example Notebook:** Introduced examples/vl_dpo_demo_gemma3.ipynb, demonstrating multimodal DPO fine-tuning with image-conditioned  prompts using multimodal gemma3
- **Gemma3 Tokenizer Adapter:** Extended TokenizerAdapter to support Hugging Face Processor objects
- **VLM Sampler and Utils:**
   1. Added VLMSampler for multimodal generation (tunix/generate/vlm_sampler.py)
   2. Added preprocess_image() utility in tunix/generate/utils.py for SigLIP/CLIP-style normalization.
- **DPO Pipeline Updates:** Modified tunix/sft/dpo/dpo_trainer.py to handle multimodal data ({“text”, “image”} prompts) and propagate image tensors through training.

@Tianjiao-Yu led this effort and @jxiong21029 contributed to the Gemma3 integration. Please also mention @Tianjiao-Yu if you have any questions/comments/feedback.

**Colab Notebook**
vl_dpo_demo_gemma3.ipynb

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by abheesht17** [comment]
@immuntasir - let me know when this is ready for review!

**Comment by immuntasir** [comment]
> @immuntasir - let me know when this is ready for review!

@Tianjiao-Yu confirmed that this is ready for review.

File: examples/dpo_demo_gemma3.ipynb
Comment (immuntasir) [inline examples/dpo_demo_gemma3.ipynb:1]: @Tianjiao-Yu I think this should be removed from this PR.

File: examples/vl_dpo_demo_gemma3.ipynb [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "79e6e5f1"
+      },
+      "source": [
+        "# Fine-tuning a Visual Language Model (VLM) using DPO\n",
+        "\n",
+        "This notebook demonstrates how to fine-tune a Visual Language Model (VLM), specifically the Gemma 3-1B-it model, using the Direct Preference Optimization (DPO) algorithm.\n",
```
Comment (abheesht17) [inline examples/vl_dpo_demo_gemma3.ipynb]: > Gemma 3-1B-it model

This is a text-only model though. 4B onwards are VLMs

File: examples/vl_dpo_demo_gemma3.ipynb [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+        "The key steps involved are:\n",
+        "\n",
+        "1.  **Setup and Installations**: Install necessary libraries and dependencies.\n",
+        "2.  **Model Loading**: Load the pre-trained Gemma 3-1B-it model.\n",
+        "3.  **LoRA Application**: Apply Low-Rank Adaptation (LoRA) to the model for efficient fine-tuning.\n",
+        "4.  **Data Loading and Preprocessing**: Load the RLAIF-V dataset and preprocess it for VLM training, including handling images and tokenizing text.\n",
+        "5.  **DPO Training**: Set up and run the DPO training loop to fine-tune the model based on preference data (chosen and rejected responses).\n",
+        "6.  **Logging and Visualization**: Log training metrics and visualize the training progress.\n",
+        "\n",
+        "The goal is to train the VLM to better align with human preferences by optimizing directly on pairs of preferred and dispreferred responses."
```
Comment (abheesht17) [inline examples/vl_dpo_demo_gemma3.ipynb]: > preferred and dispreferred

Is this common terminology? Or do we use preferred/unpreferred OR chosen/rejected?

File: tunix/sft/dpo/dpo_trainer.py [hunk 3]
```diff
@@ -25,27 +25,30 @@
...
+ImageType = np.ndarray | jax.Array | Image.Image
+
 
 @flax.struct.dataclass(frozen=True)
 class DataInput:
   """Training data input for DPO.
 
   This can be used when inputs are raw strings. Tokenization, padding and
-  preprocessing is taken care of by `DPOTrainer`.
+  preprocessing is taken care of by `DpoTrainer`.
```
Comment (abheesht17) [inline tunix/sft/dpo/dpo_trainer.py]: Revert?

File: examples/dpo_demo_gemma3.ipynb
Comment (abheesht17) [reply examples/dpo_demo_gemma3.ipynb:1]: +1

File: tunix/generate/tokenizer_adapter.py [hunk 4]
```diff
@@ -55,11 +57,16 @@ def __init__(self, tokenizer: Any):
...
-  def encode(self, text: str, **kwargs) -> list[int]:
+  def encode(self, text: str, **kwargs) -> list[int] | tuple[list[int], Any]:
     if self._tokenizer_type == TokenizerType.SP:
       return self._tokenizer.EncodeAsIds(text, **kwargs)
     elif self._tokenizer_type == TokenizerType.HF:
       return self._tokenizer.encode(text, **kwargs)
+    elif self._tokenizer_type == TokenizerType.HFP:
+      inputs = self._tokenizer(text=text, **kwargs)
+      if 'images' in kwargs:
+        return inputs['input_ids'], inputs['pixel_values']
```
Comment (abheesht17) [inline tunix/generate/tokenizer_adapter.py]: Better to return a dictionary here rather than a tuple (in case we add more modalities later)?

File: tunix/generate/tokenizer_adapter.py [hunk 5]
```diff
@@ -18,16 +18,16 @@
...
-
 import sentencepiece as spm
+import transformers
+from etils import epath
 
 
 class TokenizerType(enum.Enum):
   SP: str = 'sp'  # sentencepiece tokenizer
   HF: str = 'hf'  # huggingface tokenizer
+  HFP: str = 'hfp'  # huggingface processor
```
Comment (abheesht17) [inline tunix/generate/tokenizer_adapter.py]: Is the only difference between these two that the processor can take images, and other modalities too? If yes, do you think we should just use HF processor everywhere (and remove HF tokeniser)?

Because if `processor(text)` works, we can just use processor everywhere

File: tunix/generate/utils.py [hunk 6]
```diff
@@ -19,13 +19,18 @@
...
 
 from flax import nnx
 import jax
 from jax import lax
+import jax.image as jimg
 import jax.numpy as jnp
 
+# Defaults compatible with CLIP / many SigLIP configs; override if needed.
+_CLIP_MEAN = jnp.array([0.48145466, 0.4578275, 0.40821073], dtype=jnp.float32)
+_CLIP_STD = jnp.array([0.26862954, 0.26130258, 0.27577711], dtype=jnp.float32)
```
Comment (abheesht17) [inline tunix/generate/utils.py]: Do you think we can move it to `models/siglip`?

File: tunix/generate/utils.py [hunk 7]
```diff
@@ -723,3 +726,44 @@ def verify_state_closeness(golden_state, state, atol=1e-2):
...
+
+
+def preprocess_image(
+    images_uint8: jnp.ndarray,
+    image_size: int,
+    *,
+    mean: Iterable[float] = _CLIP_MEAN,
+    std: Iterable[float] = _CLIP_STD,
+) -> jnp.ndarray:
+  """Resize + normalize images for SigLIP.
```
Comment (abheesht17) [inline tunix/generate/utils.py]: Just SigLIP? Does it not work for other vision models? In `generate/utils.py`, we should have generic functions (as much as possible)

File: tunix/models/gemma3/model.py [hunk 8]
```diff
@@ -845,8 +907,29 @@ def __call__(
       predicted_logits: output logits predicted by the model
       new_cache: updated cache if the input cache is not None, None elsewhere.
     """
+
     new_cache = None if cache is None else {}
-    x = self.embedder.encode(last_tokens)
+
+    if self.config.multimodal:
+      assert pixel_values is not None
+      image_mask = last_tokens == 262144  # 262144: <image_soft_token>
```
Comment (abheesht17) [inline tunix/models/gemma3/model.py]: Better to define this somewhere instead of hardcoding

**Comment by jxiong21029** [comment]
Oh, I didn't mean to request so many reviews. Not sure how that happened. Maybe from the CLA failing?

File: tunix/generate/tokenizer_adapter.py (hunk 5 above)
Comment (jxiong21029) [reply tunix/generate/tokenizer_adapter.py]: I don't think every tokenizer has an associated processor definition, so it probably makes sense to have both.

**Comment by jxiong21029** [comment]
Rebased to fix emails for CLA.

**Comment by Tianjiao-Yu** [comment]
Hi, Abheesht, I made a small update to switch the demo to use Gemma-3-4B-IT, which is a true VLM, instead of Gemma-3-1B-IT. Could you please take a quick look when you have a moment? After that, I’ll proceed with resolving the remaining merge conflicts.  Thanks!

**Comment by ridcl** [comment]
Hi guys! First of all, thanks for adding multimodal support - it works great in my use case, and I can't wait to see it merged!

I noticed that currently number of images is restricted to one per input, i.e. `pixel_values` is assumed to be of size `[B, H, W, C]`, while original Gemma implementation actually allows inputs of size `[B, N, H, W, C]`. More precisely, in the original Gemma implementation, there are two versions of `patchiy_images`:

1. [gemma.mutlimodel.image.patchify_images](https://github.com/google-deepmind/gemma/blob/b38d84803b7336aed68a3e1c2c930c20df62a4e2/gemma/multimodal/image.py#L89) with signature (legacy?):

```python
def patchify_images(
    images: typing.Float["B H W C"],
    patch_size: int = _DEFAULT_PATCH_SIZE,
    padding: str = "VALID",
) -> typing.Float["P D"]: ...
```
2. [gemma.multimodal.vision.patchify_images](https://github.com/google-deepmind/gemma/blob/b38d84803b7336aed68a3e1c2c930c20df62a4e2/gemma/multimodal/vision.py#L296) with signature (actually [used in transformer](https://github.com/google-deepmind/gemma/blob/b38d84803b7336aed68a3e1c2c930c20df62a4e2/gemma/gm/nn/_transformer.py#L412)):
```python
def patchify_images(self, images: Float["*B H W C"]) -> Float["*B P D"]: ...
```

From a quick test, I think to add support for multiple images here it should be enough to add batch dimension in a few place like `PatchEmbed.__call__()`:

```python
  @jax.named_scope("patch_embed")
  def __call__(self, x: jaxtyping.Array) -> jaxtyping.Array:
    # x: [B,H,W,3] -> conv -> [B,H/P,W/P,D] -> [B,N,D]
    x = self.proj(x)
    *b, h, w, d = x.shape                       # was:  b, h, w, d = x.shape
    x = x.reshape(*b, h * w, d)             # was: x = x.reshape(b, h * w, d)
    x = shard(x, self.cfg.shd_config.act_bnd)
    return x
```
But my local branch diverged significantly from this PR, so I can't really test the full fix just yet.

**Comment by abheesht17** [comment]
> Hi, Abheesht, I made a small update to switch the demo to use Gemma-3-4B-IT, which is a true VLM, instead of Gemma-3-1B-IT. Could you please take a quick look when you have a moment? After that, I’ll proceed with resolving the remaining merge conflicts. Thanks!

Looks good. Could you please give me edit access to this branch? I'll resolve merge conflicts and make a few changes (especially regarding the multiple images point). Thanks!

**Comment by ridcl** [comment]
@abheesht17 For the multiple image support, you may want to look at [this commit](https://github.com/ridcl/tunix/commit/0b657ac7d21a32f012fb4c382af683554fa95870) as a reference point (mostly files `tunix/models/gemma3/model.py`, `tunix/models/siglip/model.py` and `tunix/models/siglip/preprocess.py`, the rest should be formatting). 

Another change that you might be interested in is [saving LoRA params for multimodal Gemma](https://github.com/ridcl/tunix/commit/e16ce11886bdae7b96e77f5a29d3dbf8941a218c). Alternatively, I can create a clear pull request for it after the current PR is merged.

File: tunix/models/siglip/preprocess.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+import jax.image as jimg
+import jax.numpy as jnp
+
+# Defaults compatible with CLIP / many SigLIP configs; override if needed.
+_CLIP_MEAN = jnp.array([0.48145466, 0.4578275, 0.40821073], dtype=jnp.float32)
+_CLIP_STD = jnp.array([0.26862954, 0.26130258, 0.27577711], dtype=jnp.float32)
+
+
+def preprocess(
```
Comment (abheesht17) [inline tunix/models/siglip/preprocess.py]: I don't see this function being used anywhere

File: examples/vl_dpo_gemma3.ipynb [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "wT6HC4URKnaQ"
+      },
+      "outputs": [],
+      "source": [
+        "gemma_tokenizer = tokenizer_lib.Tokenizer(tokenizer_path=GEMMA_TOKENIZER_PATH)"
```
Comment (abheesht17) [inline examples/vl_dpo_gemma3.ipynb:224]: Why can we not use the HF processor directly?

File: examples/vl_dpo_demo_gemma3.ipynb [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+      "metadata": {
+        "id": "McqkpUNbKnja"
+      },
+      "outputs": [],
+      "source": [
+        "MODEL_CP_PATH = local_model_path\n",
+        "\n",
+        "model_config = gemma3_model_lib.ModelConfig.gemma3_4b()\n",
+        "model_config = dataclasses.replace(\n",
+        "    model_config, multimodal=True, num_embed=262208\n",
```
Comment (abheesht17) [inline examples/vl_dpo_demo_gemma3.ipynb]: Why don't we just expose multimodal as an arg in `gemma3_model_lib.ModelConfig.gemma3_4b(multimodal=True)`?

File: tunix/models/gemma3/model.py [hunk 12]
```diff
@@ -927,18 +1001,26 @@ def __call__(
       positions: jaxtyping.Array,  # [B, L]
       cache: Cache | None,  # (sequence length L')
       attention_mask: jaxtyping.Array,  # [B, L, L']
```
Comment (abheesht17) [inline tunix/models/gemma3/model.py]: Gemma 3 is supposed to have bidirectional attention for image tokens, but I don't see that here, or in the VLM DPO notebook.
